ex4/VendingDP.Lib/ItemDetectors.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace VendingDP.Lib
+{
+    public static class ItemDetectors
+    {
+        public static CombinatedProduct DetectProduct(string productName) 
+        {
+            // If everything goes as planned, a prepared combination product is requested:
+            try
+            {
+                List<object> preparedCombinationInstructions = MenuLoader.preparedCombinations[productName];
+                string basicProductName = (string)preparedCombinationInstructions[0];
+                CombinatedProduct combinatedProduct = new CombinatedProduct(DetectBasicProduct(basicProductName), CombinatedProductStatus.Made);
+                MakeCombinatedProduct(combinatedProduct, preparedCombinationInstructions);
+                return combinatedProduct;
+            }
+            // Product name not in the prepared combination products:
+            catch (KeyNotFoundException)
+            {
+                BasicProduct basicProduct = DetectBasicProduct(productName);
+                CombinatedProduct combinatedProduct = new CombinatedProduct(basicProduct, CombinatedProductStatus.InMaking);
+                return combinatedProduct;
+            }
```

Comment:
It confusing. Why not check if you getting something from `MenuLoader`, and if not, doing it. In such case you assume the next 3 lines does not throw `KeyNotFoundException` which not always true.

Reply:
Resolved